src/foundry/common/grid/base.d.mts

```diff
@@ -0,0 +1,612 @@
+import type { InexactPartial } from "../../../types/utils.d.mts";
+
+/**
+ * The base grid class.
+ */
+declare abstract class BaseGrid {
+  /** The base grid constructor. */
+  constructor(config: BaseGrid.Configuration);
+
+  size: number;
+
+  sizeX: number;
+
+  sizeY: number;
+
+  distance: number;
+
+  units: string;
+
+  style: string;
+
+  thickness: number;
+
+  color: Color;
+
+  alpha: number;
+
+  /**
+   * The grid type (see {@link CONST.GRID_TYPES}).
+   */
+  type: number;
+
+  get isGridless(): boolean;
+
+  get isSquare(): boolean;
+
+  get isHexagonal(): boolean;
+
+  /**
+   * Calculate the total size of the canvas with padding applied, as well as the top-left coordinates of the inner
+   * rectangle that houses the scene.
+   * @param sceneWidth  - The width of the scene.
+   * @param sceneHeight - The height of the scene.
+   * @param padding     - The percentage of padding.
+   */
+  abstract calculateDimensions(sceneWidth: number, sceneHeight: number, padding: number): BaseGrid.Dimensions;
+
+  /**
+   * Returns the offset of the grid space corresponding to the given coordinates.
+   * @param coords -The coordinates
+   * @returns The offset
+   */
+  abstract getOffset(coords: BaseGrid.Coordinates): BaseGrid.Offset;
+
+  /**
+   * Returns the smallest possible range containing the offsets of all grid spaces that intersect the given bounds.
+   * If the bounds are empty (non-positive width or height), then the offset range is empty.
+   * @example
+   * ```js
+   * const [i0, j0, i1, j1] = grid.getOffsetRange(bounds);
+   * for ( let i = i0; i < i1; i++ ) {
+   *   for ( let j = j0; j < j1; j++ ) {
+   *     const offset = {i, j};
+   *     // ...
+   *   }
+   * }
+   * ```
+   * @param bounds - The bounds
+   * @returns The offset range
+   */
+  abstract getOffsetRange(bounds: Rectangle): { i0: number; j0: number; i1: number; j1: number };
```

Comment:
Consider moving the return type into an easily reusable type like `BaseGrid.OffsetRange`.